kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/partition/DynamicPartitioner.java

```diff
@@ -0,0 +1,256 @@
+          establishedWriteAll(table,
+                              elements.subList(currentIndex, elements.size()),
+                              topic,
+                              schemas);
+          return;
+        }
+        //  increase the batch size if: we actually did just test that batchSize
+        if (!(currentPartition.size() < currentBatchSize)) {
+          increaseBatchSize();
+        }
+      } catch (BigQueryException exception) {
+        // failure
+        if (isPartitioningError(exception)) {
+          decreaseBatchSize();
+          // if we've had at least 1 successful call we'll assume this is a good batch size.
+          if (successfulCallCount > 0) {
+            seeking = false; // case 1
+            logger.debug("Best batch size found (error if higher): {}", currentBatchSize);
+            establishedWriteAll(table,
+                                elements.subList(currentIndex, elements.size()),
+                                topic,
+                                schemas);
+            return;
+          }
+        } else {
+          throw new BigQueryConnectException(
+              String.format("Failed to write to BigQuery table %s", table),
+              exception);
+        }
+      }
+    }
+    // if we finished all this with only one successful call...
+    if (successfulCallCount ==  1) {
+      // then we are done finding a batch size.
+      seeking = false; // case 2
+      // decrease batch size back to the size that we actually checked.
+      decreaseBatchSize();
+      logger.debug("Best batch size found (all elements): {}", currentBatchSize);
+    }
+  }
+
+  /**
+   * writeAll intended for a generally stable currentBatchSize.
+   *
+   * <p>Write all the given elements to BigQuery in the smallest possible number of batches, as
+   * evenly as possible.
+   *
+   * <p>If we encounter a partitioning-related error, we will immediately lower the batch size and
+   * try again.
+   *
+   * <p>Every {@link #CONT_SUCCESS_COUNT_BUMP} calls, if there have been no errors, we will bump up
+   * the batch size.
+   */
+  private void establishedWriteAll(TableId table,
+                                   List<InsertAllRequest.RowToInsert> elements,
+                                   String topic,
+                                   Set<Schema> schemas) throws BigQueryConnectException,
+                                                               InterruptedException {
+    int currentIndex = 0;
+    while (currentIndex < elements.size()) {
+      try {
+        // very similar to equalPartitioner
+        // handle case where no partitioning is necessary:
+        if (elements.size() <= currentBatchSize) {
+          writer.writeRows(table, elements, topic, schemas);
+          currentIndex = elements.size();
+          return;
+        }
+
+        int numPartitions = (int)Math.ceil(elements.size() / (currentBatchSize * 1.0));
```

Comment:
Not sure how this impacts things, but one observation is that these variables are all calculated off of `elements.size()`. If there's a write failure due to size, and decreaseBatchSize is decreased, there will be a bit of skew going on, since currentIndex won't be zero, and you'll be changing batch size.

I think you might want to change this logic to calculate (at least) numPartitions based off of elements.size() - currentIndex, and think through how this impacts things.